src/lib/profiles/security/WeaveCertProvisioning.cpp

```diff
@@ -0,0 +1,767 @@
+    // Even though it happens redundantly, we do it here to ensure it happens at the right point.
+    mBinding->SetProtocolLayerCallback(HandleBindingEvent, this);
+
+    // If the binding is ready ...
+    if (mBinding->IsReady())
+    {
+        // Allocate and initialize a new exchange context.
+        err = mBinding->NewExchangeContext(mEC);
+        SuccessOrExit(err);
+        mEC->AppState = this;
+        mEC->OnMessageReceived = HandleResponse;
+        mEC->OnResponseTimeout = HandleResponseTimeout;
+        mEC->OnKeyError = HandleKeyError;
+        mEC->OnConnectionClosed = HandleConnectionClosed;
+#if WEAVE_CONFIG_ENABLE_RELIABLE_MESSAGING
+        mEC->OnAckRcvd = HandleAckRcvd;
+        mEC->OnSendError = HandleSendError;
+#endif // WEAVE_CONFIG_ENABLE_RELIABLE_MESSAGING
+
+        // Allocate buffer.
+        msgBuf = PacketBuffer::New();
+        VerifyOrExit(msgBuf != NULL, err = WEAVE_ERROR_NO_MEMORY);
+
+        // Generate a GetCertificateRequest message.
+        err = GenerateGetCertificateRequest(msgBuf, mReqType, mDoManufAttest);
+        SuccessOrExit(err);
+
+        // Enter the RequestInProgress state.
+        mState = kState_RequestInProgress;
+
+        // Send a GetCertificateRequest message to the peer.
+        err = mEC->SendMessage(kWeaveProfile_Security, kMsgType_GetCertificateRequest, msgBuf, ExchangeContext::kSendFlag_ExpectResponse);
+        msgBuf = NULL;
+        SuccessOrExit(err);
+    }
+
+    // Otherwise, if the binding needs to be prepared...
+    else
+    {
+        // Enter the PreparingBinding state. Once binding preparation is complete, the binding will call back to
+        // the WeaveCertProvEngine, which will result in SendGetCertificateRequest() being called again.
+        mState = kState_PreparingBinding;
+
+        // If the binding is in a state where it can be prepared, ask the application to prepare it by delivering a
+        // PrepareRequested API event via the binding's callback.  At some point the binding will call back to
+        // WeaveCertProvEngine::HandleBindingEvent() signaling that preparation has completed (successfully or otherwise).
+        // Note that this callback *may* happen synchronously within the RequestPrepare() method, in which case this
+        // method (SendGetCertificateRequest()) will recurse.
+        if (mBinding->CanBePrepared())
+        {
+            err = mBinding->RequestPrepare();
+            SuccessOrExit(err);
+        }
+
+        // Fail if the binding is in some state, other than Ready, that doesn't allow it to be prepared.
+        else
+        {
+            VerifyOrExit(mBinding->IsPreparing(), err = WEAVE_ERROR_INCORRECT_STATE);
+        }
+    }
+
+exit:
+    // Free the payload buffer if it hasn't been sent.
+    if (msgBuf != NULL)
+        PacketBuffer::Free(msgBuf);
+
+    if (err != WEAVE_NO_ERROR)
+    {
+        // Enter Idle state.
+        mState = kState_Idle;
```

Comment:
I expected a call to HandleRequestDone() here.  Without one, the EC will be leaked.

Reply:
done